gpkit/tests/t_examples.py

```diff
@@ -95,42 +96,34 @@ def __exit__(self, *args):
         gpkit.settings["installed_solvers"] = SOLVERS
 
 
-class NullFile(object):
-    def write(self, string):
-        pass
-
-    def close(self):
-        pass
-
+def logged_example_testcase(name,
+                            imported=IMPORTED_EXAMPLES,
+                            exampledir=EXAMPLE_DIR):
+    """Returns a method for attaching to a unittest.TestCase that imports
+    or reloads module 'name' and stores in imported[name].
+    Runs top-level code, which is typically a docs example, in the process.
 
-class StdoutCaptured(object):
-    def __init__(self, logfilename=None):
-        self.logfilename = logfilename
-
-    def __enter__(self):
-        self.original_stdout = sys.stdout
-        if self.logfilename:
-            filepath = EXAMPLE_DIR+os.sep+"%s_output.txt" % self.logfilename
-            logfile = open(filepath, "w")
-        else:
-            logfile = NullFile()
-        sys.stdout = logfile
-
-    def __exit__(self, *args):
-        sys.stdout.close()
-        sys.stdout = self.original_stdout
+    Returns a method.
+    """
+    def test(self):
+        filepath = ("".join([exampledir,
+                             os.sep,
+                             "%s_output.txt" % name])
+                    if name not in imported else None)
+        with StdoutCaptured(logfilepath=filepath):
+            if name not in imported:
+                imported[name] = importlib.import_module(name)
+            else:
+                reload(imported[name])
+        getattr(self, name)(imported[name])
```

Comment:
Yup, exactly! This is where the new test, after import, runs the test functions that were written and then moved (`getattr(self, name)`) and passes them the imported module of the appropriate name (`imported[name]`).